hudi/src/main/scala/org/apache/spark/sql/delta/hudi/HudiConverter.scala

```diff
@@ -0,0 +1,336 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.hudi
+
+import org.apache.commons.lang3.exception.ExceptionUtils
+import org.apache.hudi.common.model.{HoodieCommitMetadata, HoodieReplaceCommitMetadata}
+import org.apache.hudi.common.table.HoodieTableMetaClient
+import org.apache.hudi.common.table.timeline.{HoodieInstant, HoodieTimeline}
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.catalog.CatalogTable
+import org.apache.spark.sql.delta.actions.Action
+import org.apache.spark.sql.delta.hooks.HudiConverterHook
+import org.apache.spark.sql.delta.hudi.HudiTransactionUtils._
+import org.apache.spark.sql.delta.metering.DeltaLogging
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+import org.apache.spark.sql.delta._
+
+import java.io.{IOException, UncheckedIOException}
+import java.util.concurrent.atomic.AtomicReference
+import javax.annotation.concurrent.GuardedBy
+import scala.collection.JavaConverters._
+import scala.util.control.NonFatal
+
+object HudiConverter {
+  /**
+   * Property to be set in translated Hudi commit metadata.
+   * Indicates the delta commit version # that it corresponds to.
+   */
+  val DELTA_VERSION_PROPERTY = "delta-version"
+
+  /**
+   * Property to be set in translated Hudi commit metadata.
+   * Indicates the timestamp (milliseconds) of the delta commit that it corresponds to.
+   */
+  val DELTA_TIMESTAMP_PROPERTY = "delta-timestamp"
+}
+
+/**
+ * This class manages the transformation of delta snapshots into their Hudi equivalent.
+ */
+class HudiConverter(spark: SparkSession)
+    extends UniversalFormatConverter(spark)
+    with DeltaLogging {
+
+  // Save an atomic reference of the snapshot being converted, and the txn that triggered
+  // resulted in the specified snapshot
+  protected val currentConversion =
+    new AtomicReference[(Snapshot, OptimisticTransactionImpl)]()
+  protected val standbyConversion =
+    new AtomicReference[(Snapshot, OptimisticTransactionImpl)]()
+
+  // Whether our async converter thread is active. We may already have an alive thread that is
+  // about to shutdown, but in such cases this value should return false.
+  @GuardedBy("asyncThreadLock")
+  private var asyncConverterThreadActive: Boolean = false
+  private val asyncThreadLock = new Object
+
+  /**
+   * Enqueue the specified snapshot to be converted to Hudi. This will start an async
+   * job to run the conversion, unless there already is an async conversion running for
+   * this table. In that case, it will queue up the provided snapshot to be run after
+   * the existing job completes.
+   * Note that if there is another snapshot already queued, the previous snapshot will get
+   * removed from the wait queue. Only one snapshot is queued at any point of time.
+   *
+   */
+  override def enqueueSnapshotForConversion(
```

Comment:
Ideally we want to consolidate all these async logic with iceberg conversion... I will take that part and I am not asking for a TODO from this PR.. .